File: examples/rust/src/bin/accepted_transactions.rs
use std::sync::Arc;
use std::time::Duration;

use kaspa_rpc_core::{api::rpc::RpcApi, RpcDataVerbosityLevel};
use kaspa_wrpc_client::{
    client::{ConnectOptions, ConnectStrategy},
    prelude::{NetworkId, NetworkType},
    KaspaRpcClient, Resolver, WrpcEncoding,
};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let network_id = NetworkId::new(NetworkType::Mainnet);
    let rpc_url: Option<&str> = None;
    let start_hash: Option<&str> = None;

    let client = Arc::new(KaspaRpcClient::new_with_args(
        WrpcEncoding::Borsh,
        rpc_url,
        if rpc_url.is_some() {
            None
        } else {
            Some(Resolver::default())
        },
        Some(network_id),
        None,
    )?);

    let options = ConnectOptions {
        block_async_connect: true,
        connect_timeout: Some(Duration::from_millis(5_000)),
        strategy: ConnectStrategy::Fallback,
        ..Default::default()
    };

    client.connect(Some(options)).await?;

    let dag_info = client.get_block_dag_info().await?;
    let start_hash = if let Some(hash) = start_hash {
        hash.parse()?
    } else {
        dag_info.pruning_point_hash
    };

    let response = client
        .get_virtual_chain_from_block_v2(start_hash, Some(RpcDataVerbosityLevel::High), Some(10))
        .await?;

    // rollback previously ingested txs that are tied to no longer accepted blocks
    response
        .removed_chain_block_hashes
        .iter()
        .for_each(|removed_block_hash| {
            // Here: handle rollback
            println!("block {} is no longer accepted", removed_block_hash);
        });

    for block in response.chain_block_accepted_transactions.iter() {
        for tx in &block.accepted_transactions {
            println!(
                "{:?} {:?} {:?}",
                block.chain_block_header.hash,
                tx.verbose_data
                    .as_ref()
                    .and_then(|verbose| verbose.transaction_id),
                tx.payload
            );
        }
    }

    let next_checkpoint = response
        .added_chain_block_hashes
        .last()
        .copied()
        .unwrap_or(start_hash);
    println!("Next checkpoint: {next_checkpoint}");

    client.disconnect().await?;

    Ok(())
}

File: examples/node/accepted-transactions.js
globalThis.WebSocket = require("websocket").w3cwebsocket;

const kaspa = require("../kaspa-wasm32-sdk/nodejs/kaspa");
const { RpcClient, Resolver, Encoding } = kaspa;

const networkId = "mainnet";
const rpcUrl = undefined;
const startHash = undefined;

(async () => {
  const rpc = new RpcClient({
    networkId,
    encoding: Encoding.Borsh,
    ...(rpcUrl ? { url: rpcUrl } : { resolver: new Resolver() }),
  });

  try {
    await rpc.connect();

    const { pruningPointHash } = await rpc.getBlockDagInfo();

    const response = await rpc.getVirtualChainFromBlockV2({
      startHash: startHash ?? pruningPointHash,
      minConfirmationCount: 10,
      dataVerbosityLevel: "High",
    });

    // rollback previously ingested txs that are tied to no longer accepted blocks
    for (const removedBlockHash of response.removedChainBlockHashes) {
      // Here: handle rollback
      console.log(`block ${removedBlockHash} is no longer accepted`);
    }

    response.removedChainBlockHashes;

    for (const block of response.chainBlockAcceptedTransactions) {
      for (const tx of block.acceptedTransactions) {
        console.log(
          block.chainBlockHeader.hash,
          tx.verboseData?.transactionId,
          tx.payload,
        );
      }
    }

    const nextCheckpoint =
      response.addedChainBlockHashes[
        response.addedChainBlockHashes.length - 1
      ] ??
      startHash ??
      pruningPointHash;

    console.log("Next checkpoint:", nextCheckpoint);
  } finally {
    await rpc.disconnect().catch(() => {});
  }
})().catch((error) => {
  console.error(error);
  process.exit(1);
});
